lua_examples/luaOTA/luaOTAserver.lua

```diff
@@ -0,0 +1,240 @@
+local function main ()
+  local server     = assert(socket.bind("*", ESPport))
+  local ip, port   = server:getsockname()
+
+  print("Lua OTA service listening on  %s:%u\n After connecting, the ESP timeout is %u s" 
+     % {ip, port, ESPtimeout})
+
+  -- Main loop forever waiting for ESP clients then processing each request ------------
+
+
+  while true do
+    local esp = server:accept()           -- wait for ESP connection 
+    esp:settimeout(ESPtimeout)            -- set session timeout
+    -- receive the opening request 
+    local config = receive_and_parse(esp)
+    if config and config.a == "HI" then 
+      print ("Processing provision check from ESP-"..config.id)
+      local inventory, fingerprint = get_inventory(src_dir, config.id)
+      -- Process the ESP request
+      if config.chk and config.chk == fingerprint then
+        send_command(esp, {r = "OK!"})       -- no update so send_command with OK
+        esp:receive("*l")                    -- dummy receive to allow client to close
+      else
+        local status, msg = pcall(provision, esp, config, inventory, fingerprint)
+        if not status then print (msg) end
+  
+    end
+    end
+    pcall(esp.close, esp)
+    print ("Provisioning complete")
+  end
+end
+
+-- Local Function Implementations ------------------------------------------------------
+
+-- Enumerate the sources directory and load the relevent inventory
+------------------------------------------------------------------
+get_inventory = function(dir, cpuid)
+  if (not dir or lfs.attributes(dir).mode ~= "directory") then
+    error("Cannot open directory, aborting %s" % arg[0], 0)
+  end
+
+  -- Load the CPU's (or the default) inventory 
+  local invtype, inventory = "custom", read_file("%s/ESP-%s.json" % {dir, cpuid})  
+  if not inventory then
+    invtype, inventory = "default", read_file(dir .. "/default.json")
+  end
+
+  -- tolerate and remove whitespace formatting, then decode
+  inventory = (inventory or ""):gsub("[ \t]*\n[ \t]*","")
+  inventory = inventory:gsub("[ \t]*:[ \t]*",":")
+  local ok; ok,inventory = pcall(json.decode, inventory)
+  if ok and inventory.files then
+    print( "Loading %s inventory for ESP-%s" % {invtype, cpuid})
+  else
+    error( "Invalid inventory for %s :%s" % {cpuid,inventory}, 0)
+  end
+
+  -- Calculate the current fingerprint of the inventory
+  local fp,f = {},inventory.files
+  for i= 1,#f do
+    local name, fullname = f[i], "%s/%s" % {dir, f[i]}
+    local fa = lfs.attributes(fullname)
+
+    assert(fa, "File %s is required but not in sources directory" % name)
+    fp[#fp+1] = name .. ":" .. fa.modification
+    f[i] = {name = name, mtime = fa.modification, 
+            size = fa.size, content = read_file(fullname) }
+    assert (f[i].size == #(f[i].content or ''), "File %s unreadable" % name )
+  end
+
+  assert (#f == #fp, "Aborting provisioning die to missing fies",0)
+  return inventory, md5.sumhexa(concat(fp,":"))
```

Comment:
To perform a HMAC-MD5 with no additional packages, it suffices to use
```
function hmd5(key, data)
  if key:len() > 64 then key = md5.sum(key) end
  key = key .. string.rep('\0',64-key:len())
  local ki = md5.exor(string.rep(string.char(0x36),64),key)
  local ko = md5.exor(string.rep(string.char(0x5c),64),key)
  return md5.sumhexa(ko..md5.sum(ki..data))
end
```

This function passes all the RFC2022 (https://tools.ietf.org/html/rfc2202) test vectors, as can be verified by running

```
print(hmd5(string.rep(string.char(0x0B),16), "Hi There") == "9294727a3638bb1c13f48ef8158bfc9d")
print(hmd5("Jefe", "what do ya want for nothing?") == "750c783e6ab0b503eaa86e310a5db738")
print(hmd5(string.rep(string.char(0xaa),16), string.rep(string.char(0xdd),50)) == "56be34521d144c88dbb8c733f0e8b3f6")
print(hmd5("\1\2\3\4\5\6\7\8\9\10\11\12\13\14\15\16\17\18\19\20\21\22\23\24\25",string.rep(string.char(0xcd),50)) == "697eaf0aca3a3aea3a75164746ffaa79")
print(hmd5(string.rep(string.char(0x0c),16),"Test With Truncation") == "56461ef2342edc00f9bab995690efd4c")
print(hmd5(string.rep(string.char(0xaa),80),"Test Using Larger Than Block-Size Key - Hash Key First") == "6b1ab7fe4bd7bf8f0b62e6ce61b9d0cd")
print(hmd5(string.rep(string.char(0xaa),80),"Test Using Larger Than Block-Size Key and Larger Than One Block-Size Data") == "6f630fad67cda0ee1fb1f562db3aa53e")
```